some panic error

**Self Checks**

To make sure we get to you in time, please check the following :)
- [x] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify-plugin-daemon/issues), including closed ones.
- [x] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [x] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:)
- [x] "Please do not modify this template :) and fill in all the required fields."

**Versions**
1. dify-plugin-daemon Version
2. dify-api Version

**Describe the bug**
panic error

```
"goroutine 10317701 [running]:github.com/langgenius/dify-plugin-daemon/pkg/utils/routine.Submit.func1.1.1()/app/pkg/utils/routine/pool.go:64 +0x65panic({0x2b49440?, 0x5131a10?})/usr/local/go/src/runtime/panic.go:783 +0x132bufio.(*Writer).Flush(0xc00a744ec0)/usr/local/go/src/bufio/bufio.go:645 +0x45net/http.(*response).FlushError(0xc00e269860)/usr/local/go/src/net/http/server.go:1764 +0x33net/http.(*response).Flush(0xc00ce1d200?)/usr/local/go/src/net/http/server.go:1757 +0x13github.com/gin-gonic/gin.(*responseWriter).Flush(0xc00a6ef800?)/go/pkg/mod/github.com/gin-gonic/gin@v1.11.0/response_writer.go:129 +0xa2github.com/langgenius/dify-plugin-daemon/internal/core/io_tunnel/backwards_invocation/transaction.(*serverlessTransactionWriteCloser).Flush(0xc00f7a9b00?)/app/internal/core/io_tunnel/backwards_invocation/transaction/serverless_handler.go:41 +0x13github.com/langgenius/dify-plugin-daemon/internal/core/io_tunnel/backwards_invocation/transaction.(*ServerlessTransactionWriter).Write(0xc00cdfcd50, {0x3148b9b?, 0xc01102e750?}, {0x2931940?, 0xc009f1fe00?})/app/internal/core/io_tunnel/backwards_invocation/transaction/serverless_writer.go:42 +0x9fgithub.com/langgenius/dify-plugin-daemon/internal/core/io_tunnel/backwards_invocation.(*BackwardsInvocation).WriteError(0xc0096214a0, {0x3657700?, 0xc012350d90?})/app/internal/core/io_tunnel/backwards_invocation/request.go:60 +0xefgithub.com/langgenius/dify-plugin-daemon/internal/core/io_tunnel/backwards_invocation.executeDifyInvocationLLMTask(0xc0096214a0, 0xc00cdfcdb0?)/app/internal/core/io_tunnel/backwards_invocation/task.go:361 +0x2f8github.com/langgenius/dify-plugin-daemon/internal/core/io_tunnel/backwards_invocation.genericDispatchTask[...](0xc0096214a0?, 0x325f340)/app/internal/core/io_tunnel/backwards_invocation/task.go:297 +0x9dgithub.com/langgenius/dify-plugin-daemon/internal/core/io_tunnel/backwards_invocation.init.func19(0x2c0ec80?)/app/internal/core/io_tunnel/backwards_invocation/task.go:235 +0x25github.com/langgenius/dify-plugin-daemon/internal/core/io_tunnel/backwards_invocation.dispatchDifyInvocationTask(0xc0096214a0)/app/internal/core/io_tunnel/backwards_invocation/task.go:319 +0x385github.com/langgenius/dify-plugin-daemon/internal/core/io_tunnel/backwards_invocation.InvokeDify.func1()/app/internal/core/io_tunnel/backwards_invocation/task.go:64 +0x28github.com/langgenius/dify-plugin-daemon/pkg/utils/routine.Submit.func1.1({0x3690cb8?, 0xc01597e0f0?})/app/pkg/utils/routine/pool.go:68 +0x44runtime/pprof.Do({0x3690c80?, 0x51e4f80?}, {{0xc007998720?, 0x315cc8d?, 0x29ea400?}}, 0xc00f2f9f68)/usr/local/go/src/runtime/pprof/runtime.go:51 +0x8cgithub.com/langgenius/dify-plugin-daemon/pkg/utils/routine.Submit.func1()/app/pkg/utils/routine/pool.go:58 +0x14agithub.com/panjf2000/ants/v2.(*goWorker).run.func1()/go/pkg/mod/github.com/panjf2000/ants/v2@v2.10.0/worker.go:71 +0x8acreated by github.com/panjf2000/ants/v2.(*goWorker).run in goroutine 10270098/go/pkg/mod/github.com/panjf2000/ants/v2@v2.10.0/worker.go:48 +0x5c"
```

error

```

"github.com/langgenius/dify-plugin-daemon/internal/core/control_panel.(*ControlPanel).onDebuggingRuntimeDisconnected/app/internal/core/control_panel/server_debugger.go:56github.com/langgenius/dify-plugin-daemon/internal/core/control_panel.(*DebuggingRuntimeSignal).OnRuntimeDisconnected/app/internal/core/control_panel/debugging_signal.go:27github.com/panjf2000/gnet/v2.(*eventloop).close/go/pkg/mod/github.com/panjf2000/gnet/v2@v2.5.5/eventloop_unix.go:274github.com/panjf2000/gnet/v2.(*conn).Close.func1/go/pkg/mod/github.com/panjf2000/gnet/v2@v2.5.5/connection_unix.go:509github.com/panjf2000/gnet/v2/internal/netpoll.(*Poller).Polling/go/pkg/mod/github.com/panjf2000/gnet/v2@v2.5.5/internal/netpoll/poller_epoll_default.go:150github.com/panjf2000/gnet/v2.(*eventloop).run/go/pkg/mod/github.com/panjf2000/gnet/v2@v2.5.5/reactor_default.go:87golang.org/x/sync/errgroup.(*Group).Go.func1/go/pkg/mod/golang.org/x/sync@v0.19.0/errgroup/errgroup.go:93"
--
  | log.file.pathtype tagdata type  string | "/var/log/pods/dify-plugin_dify-plugindaemon-5977d8c866-hzmtl_53c5cfa1-977b-41fb-addd-2fd9221bfd47/dify-plugindaemon/0.log"
```

test failed

```
panic: test timed out after 1m0s
	running tests:
		TestStreamBlockingWrite (55s)

goroutine 9 [running]:
testing.(*M).startAlarm.func1()
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2682 +0x345
created by time.goFunc
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/time/sleep.go:215 +0x2d

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000102c40, {0x657138?, 0xc000179b30?}, 0x6679e0)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2005 +0x485
testing.runTests.func1(0xc000102c40)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2477 +0x37
testing.tRunner(0xc000102c40, 0xc000179c70)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0xea
testing.runTests(0xc0001280f0, {0x853fc0, 0x5, 0x5}, {0x7?, 0xc00012ab80?, 0x85ba60?})
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2475 +0x4b4
testing.(*M).Run(0xc000118960)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2337 +0x63a
main.main()
	_testmain.go:55 +0x9b

goroutine 25 [chan receive]:
github.com/langgenius/dify-plugin-daemon/pkg/utils/stream.TestStreamBlockingWrite(0xc0001036c0)
	/home/runner/work/dify-plugin-daemon/dify-plugin-daemon/pkg/utils/stream/stream_test.go:128 +0x1d2
testing.tRunner(0xc0001036c0, 0x6679e0)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0xea
created by testing.(*T).Run in goroutine 1
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x465

goroutine 27 [chan receive]:
github.com/langgenius/dify-plugin-daemon/pkg/utils/stream.(*Stream[...]).Next(0x6ad2a0)
	/home/runner/work/dify-plugin-daemon/dify-plugin-daemon/pkg/utils/stream/stream.go:79 +0x14d
github.com/langgenius/dify-plugin-daemon/pkg/utils/stream.TestStreamBlockingWrite.func2()
	/home/runner/work/dify-plugin-daemon/dify-plugin-daemon/pkg/utils/stream/stream_test.go:119 +0x85
created by github.com/langgenius/dify-plugin-daemon/pkg/utils/stream.TestStreamBlockingWrite in goroutine 25
	/home/runner/work/dify-plugin-daemon/dify-plugin-daemon/pkg/utils/stream/stream_test.go:113 +0x1c6

goroutine 26 [sleep]:
time.Sleep(0x3e8)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/runtime/time.go:363 +0x165
github.com/langgenius/dify-plugin-daemon/pkg/utils/stream.TestStreamBlockingWrite.func1()
	/home/runner/work/dify-plugin-daemon/dify-plugin-daemon/pkg/utils/stream/stream_test.go:106 +0x3b
created by github.com/langgenius/dify-plugin-daemon/pkg/utils/stream.TestStreamBlockingWrite in goroutine 25
	/home/runner/work/dify-plugin-daemon/dify-plugin-daemon/pkg/utils/stream/stream_test.go:103 +0x12a
FAIL	github.com/langgenius/dify-plugin-daemon/pkg/utils/stream	60.004s
?   	github.com/langgenius/dify-plugin-daemon/pkg/utils/strings	[no test files]
?   	github.com/langgenius/dify-plugin-daemon/pkg/utils/tmpfile	[no test files]
?   	github.com/langgenius/dify-plugin-daemon/pkg/validators	[no test files]
FAIL
```

remote plugin has not init, then disconnect the server

**To Reproduce**
Steps to reproduce the behavior:
1. Go to '...'
2. Click on '....'
3. Scroll down to '....'
4. See error

**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

--- a/.github/workflows/benchmark.yml
+++ b/.github/workflows/benchmark.yml
@@ -34,7 +34,7 @@ jobs:
         env:
           UV_PATH: ${{ env.UV_PATH }}
         run: |
-          go test -run=^$ -bench=. -benchmem -benchtime=30s ./... > result.txt
+          go test -run=^$ -bench=. -benchmem -benchtime=30s ./...
 
       - name: Remove License
         run: git stash

--- a/internal/core/control_panel/server_debugger.go
+++ b/internal/core/control_panel/server_debugger.go
@@ -50,6 +50,10 @@ func (c *ControlPanel) onDebuggingRuntimeConnected(
 func (c *ControlPanel) onDebuggingRuntimeDisconnected(
 	rpr *debugging_runtime.RemotePluginRuntime,
 ) {
+	if !rpr.Initialized() {
+		return
+	}
+
 	// handle plugin disconnecting
 	pluginIdentifier, err := rpr.Identity()
 	if err != nil {

--- a/internal/core/control_panel/server_debugger_callbacks_test.go
+++ b/internal/core/control_panel/server_debugger_callbacks_test.go
@@ -1,90 +1,113 @@
 package controlpanel
 
 import (
-    "reflect"
-    "testing"
-    "unsafe"
+	"reflect"
+	"testing"
+	"unsafe"
 
-    "github.com/google/uuid"
-    "github.com/stretchr/testify/assert"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
 
-    "github.com/langgenius/dify-plugin-daemon/internal/core/debugging_runtime"
-    "github.com/langgenius/dify-plugin-daemon/internal/core/local_runtime"
-    "github.com/langgenius/dify-plugin-daemon/internal/types/app"
-    "github.com/langgenius/dify-plugin-daemon/pkg/entities/manifest_entities"
-    "github.com/langgenius/dify-plugin-daemon/pkg/entities/plugin_entities"
+	"github.com/langgenius/dify-plugin-daemon/internal/core/debugging_runtime"
+	"github.com/langgenius/dify-plugin-daemon/internal/core/local_runtime"
+	"github.com/langgenius/dify-plugin-daemon/internal/types/app"
+	"github.com/langgenius/dify-plugin-daemon/pkg/entities/manifest_entities"
+	"github.com/langgenius/dify-plugin-daemon/pkg/entities/plugin_entities"
 )
 
 type mockNotifier struct {
-    connected    bool
-    disconnected bool
+	connected    bool
+	disconnected bool
 }
 
-func (m *mockNotifier) OnLocalRuntimeStarting(plugin_entities.PluginUniqueIdentifier) {}
-func (m *mockNotifier) OnLocalRuntimeReady(*local_runtime.LocalPluginRuntime) {}
+func (m *mockNotifier) OnLocalRuntimeStarting(plugin_entities.PluginUniqueIdentifier)           {}
+func (m *mockNotifier) OnLocalRuntimeReady(*local_runtime.LocalPluginRuntime)                   {}
 func (m *mockNotifier) OnLocalRuntimeStartFailed(plugin_entities.PluginUniqueIdentifier, error) {}
-func (m *mockNotifier) OnLocalRuntimeStop(*local_runtime.LocalPluginRuntime) {}
-func (m *mockNotifier) OnLocalRuntimeStopped(*local_runtime.LocalPluginRuntime) {}
-func (m *mockNotifier) OnLocalRuntimeScaleUp(*local_runtime.LocalPluginRuntime, int32) {}
-func (m *mockNotifier) OnLocalRuntimeScaleDown(*local_runtime.LocalPluginRuntime, int32) {}
-func (m *mockNotifier) OnLocalRuntimeInstanceLog(*local_runtime.LocalPluginRuntime, *local_runtime.PluginInstance, plugin_entities.PluginLogEvent) {}
-func (m *mockNotifier) OnDebuggingRuntimeConnected(r *debugging_runtime.RemotePluginRuntime)  { m.connected = true }
-func (m *mockNotifier) OnDebuggingRuntimeDisconnected(r *debugging_runtime.RemotePluginRuntime) { m.disconnected = true }
+func (m *mockNotifier) OnLocalRuntimeStop(*local_runtime.LocalPluginRuntime)                    {}
+func (m *mockNotifier) OnLocalRuntimeStopped(*local_runtime.LocalPluginRuntime)                 {}
+func (m *mockNotifier) OnLocalRuntimeScaleUp(*local_runtime.LocalPluginRuntime, int32)          {}
+func (m *mockNotifier) OnLocalRuntimeScaleDown(*local_runtime.LocalPluginRuntime, int32)        {}
+func (m *mockNotifier) OnLocalRuntimeInstanceLog(*local_runtime.LocalPluginRuntime, *local_runtime.PluginInstance, plugin_entities.PluginLogEvent) {
+}
+func (m *mockNotifier) OnDebuggingRuntimeConnected(r *debugging_runtime.RemotePluginRuntime) {
+	m.connected = true
+}
+func (m *mockNotifier) OnDebuggingRuntimeDisconnected(r *debugging_runtime.RemotePluginRuntime) {
+	m.disconnected = true
+}
 
 // setPrivateString sets an unexported string field on a struct value via unsafe reflection.
 func setPrivateString(target any, field string, value string) {
-    rv := reflect.ValueOf(target).Elem()
-    f := rv.FieldByName(field)
-    reflect.NewAt(f.Type(), unsafe.Pointer(f.UnsafeAddr())).Elem().SetString(value)
+	rv := reflect.ValueOf(target).Elem()
+	f := rv.FieldByName(field)
+	reflect.NewAt(f.Type(), unsafe.Pointer(f.UnsafeAddr())).Elem().SetString(value)
 }
 
 func newFakeRemoteRuntime(t *testing.T, name, version string) *debugging_runtime.RemotePluginRuntime {
-    t.Helper()
+	t.Helper()
 
-    r := &debugging_runtime.RemotePluginRuntime{
-        PluginRuntime: plugin_entities.PluginRuntime{
-            Config: plugin_entities.PluginDeclaration{
-                PluginDeclarationWithoutAdvancedFields: plugin_entities.PluginDeclarationWithoutAdvancedFields{
-                    // Author is overwritten with tenantId inside Identity()
-                    Name:    name,
-                    Version: manifest_entities.Version(version),
-                },
-            },
-        },
-    }
+	r := &debugging_runtime.RemotePluginRuntime{
+		PluginRuntime: plugin_entities.PluginRuntime{
+			Config: plugin_entities.PluginDeclaration{
+				PluginDeclarationWithoutAdvancedFields: plugin_entities.PluginDeclarationWithoutAdvancedFields{
+					// Author is overwritten with tenantId inside Identity()
+					Name:    name,
+					Version: manifest_entities.Version(version),
+				},
+			},
+		},
+	}
 
-    // Provide required tenantId and checksum so Identity() succeeds
-    setPrivateString(r, "tenantId", uuid.New().String())
-    setPrivateString(r, "checksum", "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef")
+	// Provide required tenantId and checksum so Identity() succeeds
+	setPrivateString(r, "tenantId", uuid.New().String())
+	setPrivateString(r, "checksum", "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef")
+	reflect.NewAt(
+		reflect.ValueOf(r).Elem().FieldByName("initialized").Type(),
+		unsafe.Pointer(reflect.ValueOf(r).Elem().FieldByName("initialized").UnsafeAddr()),
+	).Elem().SetBool(true)
 
-    return r
+	return r
 }
 
 func TestOnDebuggingRuntimeConnectedAndDisconnected(t *testing.T) {
-    cp := NewControlPanel(&app.Config{PluginLocalLaunchingConcurrent: 1}, nil, nil, nil, nil)
+	cp := NewControlPanel(&app.Config{PluginLocalLaunchingConcurrent: 1}, nil, nil, nil, nil)
 
-    // Attach mock notifier to observe callbacks
-    mn := &mockNotifier{}
-    cp.AddNotifier(mn)
+	// Attach mock notifier to observe callbacks
+	mn := &mockNotifier{}
+	cp.AddNotifier(mn)
 
-    r := newFakeRemoteRuntime(t, "conn_test", "1.2.3")
+	r := newFakeRemoteRuntime(t, "conn_test", "1.2.3")
 
-    // Connected path
-    err := cp.onDebuggingRuntimeConnected(r)
-    assert.NoError(t, err)
+	// Connected path
+	err := cp.onDebuggingRuntimeConnected(r)
+	assert.NoError(t, err)
 
-    id, err := r.Identity()
-    assert.NoError(t, err)
+	id, err := r.Identity()
+	assert.NoError(t, err)
 
-    // Stored in runtime map and notifier called
-    _, ok := cp.debuggingPluginRuntime.Load(id)
-    assert.True(t, ok, "runtime should be stored on connect")
-    assert.True(t, mn.connected, "connected notifier should be triggered")
+	// Stored in runtime map and notifier called
+	_, ok := cp.debuggingPluginRuntime.Load(id)
+	assert.True(t, ok, "runtime should be stored on connect")
+	assert.True(t, mn.connected, "connected notifier should be triggered")
 
-    // Disconnected path
-    cp.onDebuggingRuntimeDisconnected(r)
+	// Disconnected path
+	cp.onDebuggingRuntimeDisconnected(r)
 
-    _, ok = cp.debuggingPluginRuntime.Load(id)
-    assert.False(t, ok, "runtime should be removed on disconnect")
-    assert.True(t, mn.disconnected, "disconnected notifier should be triggered")
+	_, ok = cp.debuggingPluginRuntime.Load(id)
+	assert.False(t, ok, "runtime should be removed on disconnect")
+	assert.True(t, mn.disconnected, "disconnected notifier should be triggered")
+}
+
+func TestOnDebuggingRuntimeDisconnected_IgnoresUninitializedRuntime(t *testing.T) {
+	cp := NewControlPanel(&app.Config{PluginLocalLaunchingConcurrent: 1}, nil, nil, nil, nil)
+
+	mn := &mockNotifier{}
+	cp.AddNotifier(mn)
+
+	r := &debugging_runtime.RemotePluginRuntime{}
+
+	assert.NotPanics(t, func() {
+		cp.onDebuggingRuntimeDisconnected(r)
+	})
+	assert.False(t, mn.disconnected, "disconnected notifier should not be triggered for uninitialized runtime")
 }

--- a/internal/core/debugging_runtime/type.go
+++ b/internal/core/debugging_runtime/type.go
@@ -156,6 +156,10 @@ func (r *RemotePluginRuntime) SetInstallationId(installationId string) {
 	r.installationId = installationId
 }
 
+func (r *RemotePluginRuntime) Initialized() bool {
+	return r.initialized
+}
+
 func (r *RemotePluginRuntime) Identity() (plugin_entities.PluginUniqueIdentifier, error) {
 	// FIXME: it's a little bit tricky that replace author with current tenant_id
 	// just as a flag to identify debugging plugin

--- a/internal/core/io_tunnel/backwards_invocation/transaction/serverless_handler.go
+++ b/internal/core/io_tunnel/backwards_invocation/transaction/serverless_handler.go
@@ -1,8 +1,10 @@
 package transaction
 
 import (
+	"fmt"
 	"io"
 	"net/http"
+	"sync"
 	"sync/atomic"
 	"time"
 
@@ -28,17 +30,46 @@ func NewServerlessTransactionHandler(maxTimeout time.Duration) *ServerlessTransa
 type serverlessTransactionWriteCloser struct {
 	done   chan bool
 	closed int32
+	mu     sync.Mutex
 
 	writer func([]byte) (int, error)
 	flush  func()
 }
 
-func (a *serverlessTransactionWriteCloser) Write(data []byte) (int, error) {
-	return a.writer(data)
+func (w *serverlessTransactionWriteCloser) Write(data []byte) (n int, err error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if atomic.LoadInt32(&w.closed) != 0 {
+		return 0, io.ErrClosedPipe
+	}
+
+	defer func() {
+		if recovered := recover(); recovered != nil {
+			n = 0
+			err = fmt.Errorf("serverless transaction write panic: %v", recovered)
+			_ = w.Close()
+		}
+	}()
+
+	return w.writer(data)
 }
 
-func (a *serverlessTransactionWriteCloser) Flush() {
-	a.flush()
+func (w *serverlessTransactionWriteCloser) Flush() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if atomic.LoadInt32(&w.closed) != 0 {
+		return
+	}
+
+	defer func() {
+		if recover() != nil {
+			_ = w.Close()
+		}
+	}()
+
+	w.flush()
 }
 
 func (w *serverlessTransactionWriteCloser) Close() error {
@@ -63,9 +94,6 @@ func (h *ServerlessTransactionHandler) Handle(ctx *gin.Context, sessionId string
 		ctx.Writer.Write([]byte(err.Error()))
 		return
 	}
-
-	ctx.Writer.WriteHeader(http.StatusOK)
-	ctx.Writer.Header().Set("Content-Type", "text/event-stream")
 
 	plugin_entities.ParsePluginUniversalEvent(
 		bytes,
@@ -121,6 +149,8 @@ func (h *ServerlessTransactionHandler) Handle(ctx *gin.Context, sessionId string
 				return
 			}
 
+			ctx.Writer.Header().Set("Content-Type", "text/event-stream")
+
 			// replace trace context, propagate it to gin
 			ctxRequestContext := ctx.Request.Context()
 			ctxRequestContext = log.WithTrace(ctxRequestContext, session.TraceContext)
@@ -128,8 +158,8 @@ func (h *ServerlessTransactionHandler) Handle(ctx *gin.Context, sessionId string
 			ctx.Request = ctx.Request.WithContext(ctxRequestContext)
 
 			// bind the backwards invocation
-			plugin_manager := plugin_manager.Manager()
-			session.BindBackwardsInvocation(plugin_manager.BackwardsInvocation())
+			pluginManager := plugin_manager.Manager()
+			session.BindBackwardsInvocation(pluginManager.BackwardsInvocation())
 
 			serverlessResponseWriter := NewServerlessTransactionWriter(session, writer)
 
@@ -140,8 +170,10 @@ func (h *ServerlessTransactionHandler) Handle(ctx *gin.Context, sessionId string
 				serverlessResponseWriter,
 				sessionMessage.Data,
 			); err != nil {
-				ctx.Writer.WriteHeader(http.StatusInternalServerError)
-				ctx.Writer.Write([]byte("failed to parse request"))
+				if !ctx.Writer.Written() {
+					ctx.Writer.WriteHeader(http.StatusInternalServerError)
+					ctx.Writer.Write([]byte("failed to parse request"))
+				}
 				writer.Close()
 			}
 		},

--- a/internal/core/io_tunnel/backwards_invocation/transaction/serverless_handler_test.go
+++ b/internal/core/io_tunnel/backwards_invocation/transaction/serverless_handler_test.go
@@ -3,7 +3,9 @@ package transaction
 import (
 	"bytes"
 	"encoding/json"
+	"io"
 	"net/http/httptest"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -53,6 +55,10 @@ func TestHandle_SessionNotFound_WritesErrorResponse(t *testing.T) {
 	h.Handle(ctx, "ignored")
 
 	// Assert
+	if recorder.Code != 400 {
+		t.Fatalf("expected status code 400, got %d", recorder.Code)
+	}
+
 	var resp backwards_invocation.BackwardsInvocationResponseEvent
 	if err := json.Unmarshal(recorder.Body.Bytes(), &resp); err != nil {
 		t.Fatalf("response should be JSON BackwardsInvocationResponseEvent, got error: %v, body: %s", err, recorder.Body.String())
@@ -89,5 +95,65 @@ func TestHandle_SessionNotFound_WritesErrorResponse(t *testing.T) {
 	}
 	if v, _ := m["detail"].(string); v == "" {
 		t.Fatalf("expected non-empty data.detail, got empty")
+	}
+}
+
+func TestServerlessTransactionWriteCloser_WriteRecoversFromPanic(t *testing.T) {
+	w := &serverlessTransactionWriteCloser{
+		done: make(chan bool),
+		writer: func([]byte) (int, error) {
+			panic("boom")
+		},
+		flush: func() {},
+	}
+
+	n, err := w.Write([]byte("payload"))
+	if err == nil {
+		t.Fatal("expected write error after panic")
+	}
+	if n != 0 {
+		t.Fatalf("expected zero bytes written, got %d", n)
+	}
+	if got := err.Error(); got != "serverless transaction write panic: boom" {
+		t.Fatalf("unexpected error: %q", got)
+	}
+	if atomic.LoadInt32(&w.closed) == 0 {
+		t.Fatal("expected writer to be closed after panic")
+	}
+}
+
+func TestServerlessTransactionWriteCloser_FlushRecoversFromPanic(t *testing.T) {
+	w := &serverlessTransactionWriteCloser{
+		done:   make(chan bool),
+		writer: func(data []byte) (int, error) { return len(data), nil },
+		flush: func() {
+			panic("boom")
+		},
+	}
+
+	w.Flush()
+
+	if atomic.LoadInt32(&w.closed) == 0 {
+		t.Fatal("expected writer to be closed after flush panic")
+	}
+}
+
+func TestServerlessTransactionWriteCloser_WriteAfterClose(t *testing.T) {
+	w := &serverlessTransactionWriteCloser{
+		done:   make(chan bool),
+		writer: func(data []byte) (int, error) { return len(data), nil },
+		flush:  func() {},
+	}
+
+	if err := w.Close(); err != nil {
+		t.Fatalf("close failed: %v", err)
+	}
+
+	_, err := w.Write([]byte("payload"))
+	if err == nil {
+		t.Fatal("expected closed pipe error")
+	}
+	if err != io.ErrClosedPipe {
+		t.Fatalf("expected io.ErrClosedPipe, got %v", err)
 	}
 }

--- a/pkg/entities/plugin_entities/event.go
+++ b/pkg/entities/plugin_entities/event.go
@@ -2,6 +2,7 @@ package plugin_entities
 
 import (
 	"encoding/json"
+	"fmt"
 
 	"github.com/langgenius/dify-plugin-daemon/pkg/utils/log"
 	"github.com/langgenius/dify-plugin-daemon/pkg/utils/parser"
@@ -24,13 +25,25 @@ func ParsePluginUniversalEvent(
 	errorHandler func(err string),
 	logHandler func(logEvent PluginLogEvent),
 ) {
+	defer func() {
+		if recovered := recover(); recovered != nil {
+			errMsg := fmt.Sprintf("plugin event handler panic: %v", recovered)
+			log.Error("plugin event handler panic", "error", recovered, "event_payload", string(data))
+			if errorHandler != nil {
+				errorHandler(errMsg)
+			}
+		}
+	}()
+
 	// handle event
 	event, err := parser.UnmarshalJsonBytes[PluginUniversalEvent](data)
 	if err != nil {
-		if len(data) > 1024 {
-			errorHandler(err.Error() + " status: " + statusText + " original response: " + string(data[:1024]) + "...")
-		} else {
-			errorHandler(err.Error() + " status: " + statusText + " original response: " + string(data))
+		if errorHandler != nil {
+			if len(data) > 1024 {
+				errorHandler(err.Error() + " status: " + statusText + " original response: " + string(data[:1024]) + "...")
+			} else {
+				errorHandler(err.Error() + " status: " + statusText + " original response: " + string(data))
+			}
 		}
 		return
 	}
@@ -53,11 +66,17 @@ func ParsePluginUniversalEvent(
 			}
 		}
 	case PLUGIN_EVENT_SESSION:
-		sessionHandler(sessionId, event.Data)
+		if sessionHandler != nil {
+			sessionHandler(sessionId, event.Data)
+		}
 	case PLUGIN_EVENT_ERROR:
-		errorHandler(string(event.Data))
+		if errorHandler != nil {
+			errorHandler(string(event.Data))
+		}
 	case PLUGIN_EVENT_HEARTBEAT:
-		heartbeatHandler()
+		if heartbeatHandler != nil {
+			heartbeatHandler()
+		}
 	}
 }
 

--- a/pkg/entities/plugin_entities/event_test.go
+++ b/pkg/entities/plugin_entities/event_test.go
@@ -1,0 +1,57 @@
+package plugin_entities
+
+import (
+	"encoding/json"
+	"strings"
+	"testing"
+)
+
+func TestParsePluginUniversalEvent_RecoversFromSessionHandlerPanic(t *testing.T) {
+	event := PluginUniversalEvent{
+		SessionId: "session-1",
+		Event:     PLUGIN_EVENT_SESSION,
+		Data:      json.RawMessage(`{"type":"invoke","data":{"x":1}}`),
+	}
+
+	var gotErr string
+
+	ParsePluginUniversalEvent(
+		mustMarshalEvent(t, event),
+		"",
+		func(string, []byte) {
+			panic("boom")
+		},
+		nil,
+		func(err string) {
+			gotErr = err
+		},
+		nil,
+	)
+
+	if !strings.Contains(gotErr, "plugin event handler panic: boom") {
+		t.Fatalf("expected panic to be forwarded to error handler, got %q", gotErr)
+	}
+}
+
+func TestParsePluginUniversalEvent_NilHandlersDoNotPanic(t *testing.T) {
+	cases := []PluginUniversalEvent{
+		{Event: PLUGIN_EVENT_SESSION, Data: json.RawMessage(`{"type":"invoke","data":{}}`)},
+		{Event: PLUGIN_EVENT_ERROR, Data: json.RawMessage(`"err"`)},
+		{Event: PLUGIN_EVENT_HEARTBEAT, Data: json.RawMessage(`null`)},
+	}
+
+	for _, event := range cases {
+		ParsePluginUniversalEvent(mustMarshalEvent(t, event), "", nil, nil, nil, nil)
+	}
+}
+
+func mustMarshalEvent(t *testing.T, event PluginUniversalEvent) []byte {
+	t.Helper()
+
+	data, err := json.Marshal(event)
+	if err != nil {
+		t.Fatalf("marshal event: %v", err)
+	}
+
+	return data
+}

--- a/pkg/utils/stream/stream_test.go
+++ b/pkg/utils/stream/stream_test.go
@@ -96,37 +96,42 @@ func TestStreamGeneratorWrapper(t *testing.T) {
 
 func TestStreamBlockingWrite(t *testing.T) {
 	response := NewStream[int](1)
-	response.Write(1)
+	assert.NoError(t, response.Write(1))
 
-	const numWrites = 1000000
+	writerStarted := make(chan struct{})
+	writerFinished := make(chan struct{})
 
 	go func() {
-		for i := 0; i < numWrites; i++ {
-			response.WriteBlocking(1)
-			time.Sleep(time.Microsecond)
-		}
-		response.Close()
+		close(writerStarted)
+		response.WriteBlocking(2)
+		close(writerFinished)
 	}()
 
-	received := 0
-	done := make(chan bool)
-	go func() {
-		defer func() {
-			close(done)
-		}()
-		// wait for the blocking write to happen
-		time.Sleep(1 * time.Second)
-		for response.Next() {
-			_, err := response.Read()
-			if err != nil {
-				t.Error(err)
-			}
-			received += 1
-		}
-	}()
+	<-writerStarted
 
-	<-done
-	assert.Equal(t, received, numWrites+1)
+	select {
+	case <-writerFinished:
+		t.Fatal("WriteBlocking should block while the queue is full")
+	case <-time.After(20 * time.Millisecond):
+	}
+
+	assert.True(t, response.Next())
+
+	first, err := response.Read()
+	assert.NoError(t, err)
+	assert.Equal(t, 1, first)
+
+	select {
+	case <-writerFinished:
+	case <-time.After(1 * time.Second):
+		t.Fatal("WriteBlocking did not unblock after the queue had space")
+	}
+
+	assert.True(t, response.Next())
+
+	second, err := response.Read()
+	assert.NoError(t, err)
+	assert.Equal(t, 2, second)
 }
 
 // WriteBlocking should return directly if the stream is closed

